api/v1/categories/CategoryCategoryController.php

```diff
@@ -0,0 +1,364 @@
+     */
+    public function add(Request $illuminateRequest): JsonResponse
+    {
+        return $this->saveCategory($illuminateRequest);
+    }
+
+    /**
+     * Edit an existing category by ID.
+     */
+    public function edit(Request $illuminateRequest): JsonResponse
+    {
+        return $this->saveCategory($illuminateRequest);
+    }
+
+    /**
+     * Create or update a category.
+     *
+     * Used internally to handle both new category creation and editing existing ones.
+     */
+    private function saveCategory(Request $illuminateRequest): JsonResponse
+    {
+        $context = $this->getRequest()->getContext();
+        $user = $this->getRequest()->getUser();
+        $parentId = (int)$illuminateRequest->input('parentCategoryId') ?: null;
+        $categoryId = (int)$illuminateRequest->route('categoryId') ?: null;
+
+        if ($parentId) {
+            $parent = Repo::category()->get($parentId, $context->getId());
+            if (!$parent) {
+                return response()->json(__('api.404.resourceNotFound'), Response::HTTP_NOT_FOUND);
+            }
+        }
+
+        // Get a category object to edit or create
+        if ($categoryId == null) {
+            $category = Repo::category()->dao->newDataObject();
+            $category->setContextId($context->getId());
+        } else {
+            $category = Repo::category()->get($categoryId, $context->getId());
+
+            if (!$category) {
+                return response()->json(__('api.404.resourceNotFound'), Response::HTTP_NOT_FOUND);
+            }
+
+            $parentId = $category->getParentId();
+        }
+
+        $params = $this->convertStringsToSchema(\PKP\services\PKPSchemaService::SCHEMA_CATEGORY, $illuminateRequest->input());
+        $params['contextId'] = $category->getContextId();
+        $errors = Repo::category()->validate($category, $params, $context);
+
+        if (!empty($errors)) {
+            return response()->json($errors, Response::HTTP_BAD_REQUEST);
+        }
+
+        // Set the editable properties of the category object
+        $category->setTitle($illuminateRequest->input('title'), null);
+        $category->setDescription($illuminateRequest->input('description'), null);
+        $category->setParentId($parentId);
+        $category->setPath($illuminateRequest->input('path'));
+        $category->setSortOption($illuminateRequest->input('sortOption'));
+
+        // Prevent categories from referencing themselves as their parent
+        if ($parentId && $categoryId && $$categoryId === $parentId) {
+            return response()->json(__('api.categories.400.invalidParentId'), Response::HTTP_BAD_REQUEST);
+        }
+
+        // Prevent category from being updated to have a circular parent reference
+        if ($parent && $category->getParentId() === $parent->getId() && $parent->getParentId() === $category->getId()) {
```

Comment:
Given that you're setting `$category->setParentId($parentId);` some lines above., then `$category->getParentId() === $parent->getId()` will be always `true` and could be removed.

Reply:
And as I said before, while we're doing our integrity checks here, another user might be changing the database and breaking the checks which you validated.

This `parentId` structure is bad, but I don't think it should be changed in this PR.

Besides fixing the code (visit the parents until you reach the top, and ensure a node isn't visited twice), I think it's needed to lock the involved items for update (the current item + parents).

Reply:
Thanks @jonasraoni. I think for this PR, adding the circular reference checks you suggest is a good idea, but as you mention the rest we should wait until this PR has been merged. If you have an idea for performing the circular references, could you share your suggested approach here?

For the other things you've mentioned, could you create a separate issue as a sub-issue of #11307 outlining some of the concerns you've mentioned here as well as the solutions you have in mind. Definitely seems like a good idea but it would be helpful to have this code solidified and merged first. Thanks again for taking such a thorough look!

Reply:
Something like this should be enough:

```php
public function hasCircularReference(Category $category, ?Context $context): bool
{
    for ($visited = []; $category?->parentId !== null;) {
        // An item cannot be visited twice
        if (isset($visited[$category->id])) {
            return true;
        }

        $visited[$category->id] = true;
        $category = Repo::category()->get($category->parentId, $context?->getId());
    }

    return false;
}
```

Reply:
I've created the sub-issue here: #11360

Reply:
For the sake of safety (even an intentionally broken database should be safe), a similar protection should be added to the code, on the places I've highlighted before.

To avoid duplicating code, a generic method that retrieves the parents, while doing the validation (you can update the method I've shared above to store the parents and retrieve them if there's no circular reference) sounds very okayish.